lib/core/src/math/utils.rs

```diff
@@ -38,11 +77,28 @@ mod tests {
                                 13, 14, 15, 16)";
         assert_eq!(column_major, expected);
     }
+
+    #[test]
+    fn matrix4_memory_layout() {
+        use super::Matrix4;
+        let matrix = Matrix4::<f32>::new
+            ( 1.0, 5.0,  9.0, 13.0
+            , 2.0, 6.0, 10.0, 14.0
+            , 3.0, 7.0, 11.0, 15.0
+            , 4.0, 8.0, 12.0, 16.0 );
+        let matrix = matrix.as_ref();
+        unsafe {
+            let layout = &*(matrix as *const [[f32; 4]; 4] as *const [f32; 16]);
+            for (i, n) in layout.iter().enumerate() {
+                assert_eq!((i + 1) as f32, *n);
+            }
+        }
```

Comment:
Undocumented usage of unsafe. Moreover, I assume that this was meant to check if our "flattenning method" works. But instead of using the method it copy-pastes the logic code, so in fact it doesnt check if our utils are correct. We can refactor one of the codes, and this tests will be not synchronised anymore.

Reply:
Thanks for pointing it out. I refactored this code to:

```rust
pub fn matrix4_to_array(matrix:&Matrix4<f32>) -> &[f32; 16] {
    // We need to interpret [[f32; 4]; 4] as [f32; 16] to view as Float32Array.
    // The memory layout is the same, so this operation is safe.
    unsafe {
        &*(matrix.as_ref() as *const [[f32; 4]; 4] as *const [f32; 16])
    }
}
```

Reply:
This does not solve my issue. The code which you are testing for correctness is this one `&*(matrix.as_ref() as *const [[f32; 4]; 4] as *const [f32; 16])`, right? It is copy-pasted here from its original definition point in `into_float32_array_view`. Thus if I change the definition of `into_float32_array_view` then this test will not test the correctness anymore. If you want to test if an unsafe function works correctly, write tests for `into_float32_array_view` instead, not for copy-pasted partial implementation of it. We see now that these are connected because they are in a single PR, but soon, after merge they will become 2 codes with similar implementations, and after someone refactors one of them, they can not be synchronised anymore.

Reply:
@wdanilo it was refactored into a `matrix4_to_array` function and I am using the same function in `into_float32_array_view`:

```rust
impl IntoFloat32ArrayView for Matrix4<f32> {
    unsafe fn into_float32_array_view(&self) -> Float32ArrayView<'_> {
        let matrix = matrix4_to_array(&self);
        let array = Float32Array::view(matrix);
        let phantom = PhantomData;
        Float32ArrayView { array, phantom }
    }
}
```

and in our tests

```rust
    #[test]
    fn matrix4_memory_layout() {
        use super::Matrix4;
        use super::matrix4_to_array;

        let matrix = Matrix4::<f32>::new
            ( 1.0, 5.0,  9.0, 13.0
            , 2.0, 6.0, 10.0, 14.0
            , 3.0, 7.0, 11.0, 15.0
            , 4.0, 8.0, 12.0, 16.0 );

        let layout = matrix4_to_array(&matrix);
        for (i, n) in layout.iter().enumerate() {
            assert_eq!((i + 1) as f32, *n);
        }
    }
```

So they aren't copy-pasted.

Reply:
This is definitely step in the right direction, but now we've got an unsafe function which is not marked unsafe!  The whole "safety" topic is very, very important for us and we should never accept it this way. Every unsafe function should be marked as unsafe and have safety notes. 

```rust
pub fn matrix4_to_array(matrix:&Matrix4<f32>) -> &[f32; 16] {
    // To interpret [[f32; 4]; 4] as [f32; 16].
    // The memory layout is the same, so this operation is safe.
    unsafe {
        &*(matrix.as_ref() as *const [[f32; 4]; 4] as *const [f32; 16])
    }
}
```

Reply:
@wdanilo my understanding is that the compiler isn't smart enough to know that `[[f32; 4]; 4]` is the same of `[f32; 16]`. It's just a different way to access this data.

The only way I found to achieve it right now is to use unsafe methods. But why would it be unsafe if the memory layout is the same?

@farmaazon do you have any consideration to make on this topic?

Reply:
@notdanilo Oh I'm so sorry, you are right. I was somehow sure that this function includes a unsafe view on JS memory and it is still unsafe. All good now!